app/Dockerfile

```diff
@@ -2,29 +2,60 @@
 # https://hub.docker.com/_/python
 
 # The build stage that will be used to deploy to the various environments
-# needs to be called `release` in order to integrate with the repo's
-# top-level Makefile
-FROM python:3-slim AS release
-
-# Install poetry, the package manager.
-# https://python-poetry.org
-RUN pip install poetry
+# This is what the other stages will call as their base
+FROM python:3.11-slim-bullseye AS build
 
+# Keep container packages up-to-date.
 RUN apt-get update \
     && apt-get install --no-install-recommends --yes \
     build-essential \
     libpq-dev \
     postgresql
 
+# Install poetry, the package manager - https://python-poetry.org
+# We use --no-cache-dir here because the pip caches normally, and if 
+#  we don't cache, it will reduce the overall image size
+# The version is pinned so that there is a known good configuration 
+#  that we have better control over vs a range that can be used
+RUN pip install --no-cache-dir poetry==1.3.1
+
+# setup user stuff
+ARG RUN_UID
+ARG RUN_USER
+
+# The following logic creates the RUN_USER home directory and the directory where
+#  we will be storing the application in the image. This runs when the user is not root
```

Comment:
i noticed in a few places you add an extra space in the comment e.g. `#  foo` instead of `# foo`, is that intentional?

Reply:
Yeah, I usually do that to show that it is a continuation of the previous comment. Not too sure where I picked it up, but I can remove that extra space if it is an issue

Reply:
Oh I see. The designer in me thinks it looks off but I don't have strong feelings. Indenting the first line of a paragraph is more typical in traditional articles/papers rather than a hanging indent. Like:

```
#   Lorem ipsum
# dolores foo bar
# cat dog mouse.
#   Lorem ipsum
# dolores foo bar
# cat dog mouse.
```
 In code I think it's also visually easy to read justified text with an extra line between paragraphs.
```
# Lorem ipsum
# dolores foo bar
# cat dog mouse.
#
# Lorem ipsum
# dolores foo bar
# cat dog mouse.
```
slight preference for the latter for visual consistency but it's not something that really matters at the end of the day